loopy/kernel/dependency.py

```diff
@@ -0,0 +1,208 @@
+import islpy as isl
+
+from loopy.symbolic import BatchedAccessMapMapper
+from loopy import LoopKernel
+from loopy import InstructionBase
+
+from dataclasses import dataclass
+from enum import Enum
+
+class DependencyType(Enum):
+    """An enumeration of the types of data dependencies found in a program.
+    """
+    WRITE_READ  = 0
+    READ_WRITE  = 1
+    WRITE_WRITE = 2
+
+class AccessType(Enum):
+    """An enumeration of the types of accesses made by statements in a program.
+    """
+    READ  = 0
+    WRITE = 1
+
+@dataclass(frozen=True)
+class HappensBefore: 
+    """A class representing a "happens-before" relationship between two
+    statements found in a :class:`loopy.LoopKernel`. Used to validate that a
+    given kernel transformation respects the data dependencies in a given
+    program.
+
+    .. attribute:: happens_before
+        The :attr:`id` of a :class:`loopy.InstructionBase` that depends on the
+        current :class:`loopy.InstructionBase` instance.
+
+    .. attribute:: variable_name
+        The name of the variable in a program that is causing the dependency. 
+
+    .. attribute:: relation
+        An :class:`isl.Map` representing the data dependency. The input of the
+        map is an iname tuple and the output of the map is a set of iname tuples
+        that must execute after the input.
+
+    .. attribute:: dependency_type
+        A :class:`DependencyType` of :class:`Enum` representing the dependency
+        type (write-read, read-write, write-write). 
+    """
+    
+    happens_before: str
+    variable_name: str
+    relation: isl.Map
+    dependency_type: DependencyType
```

Comment:
```suggestion
    variable_name: Optional[str]
    relation: isl.Map
    dependency_type: Optional[DependencyType]
```
(But probably better to make a data-mediated dependency a separate type, maybe a subclass.)